src/vlogger/vlogger.cpp

```diff
@@ -216,58 +242,113 @@ void printf_backtrace(void)
-static xlio_log_cb_t xlio_log_get_cb_func()
+// room for color code termination and EOL
+#define VLOGGER_STR_TERMINATION_SIZE 6
+
+// for the extreme case where we have a failure before initializing doca logger
+static void output_before_doca_logger(vlog_levels_t log_level, const char *fmt, ...)
 {
-    xlio_log_cb_t log_cb = NULL;
-    const char *const CB_STR = getenv(XLIO_LOG_CB_ENV_VAR);
-    if (!CB_STR || !*CB_STR) {
-        return NULL;
+    int len = 0;
+    char buf[VLOGGER_STR_SIZE];
+
+    // Format header
+
+    // Set color scheme
+    if (g_vlogger_log_in_colors) {
+        len +=
+            snprintf(buf + len, VLOGGER_STR_SIZE - len - 1, "%s", log_level::get_color(log_level));
     }
 
-    if (1 != sscanf(CB_STR, "%p", &log_cb)) {
-        return NULL;
+    len += snprintf(buf + len, VLOGGER_STR_SIZE - len - 1, " %s %s: ", g_vlogger_module_name,
+                    log_level::to_str(log_level));
+
+    if (len < 0) {
+        return;
+    }
+    buf[len + 1] = '\0';
+
+    // Format body
+    va_list ap;
+    va_start(ap, fmt);
+    if (fmt != NULL) {
+        len += vsnprintf(buf + len, VLOGGER_STR_SIZE - len, fmt, ap);
     }
-    return log_cb;
+    va_end(ap);
+
+    // Reset color scheme
+    if (g_vlogger_log_in_colors) {
+        // Save enough room for color code termination and EOL
+        if (len > VLOGGER_STR_SIZE - VLOGGER_STR_TERMINATION_SIZE) {
+            len = VLOGGER_STR_SIZE - VLOGGER_STR_TERMINATION_SIZE - 1;
+        }
+
+        len = snprintf(buf + len, VLOGGER_STR_TERMINATION_SIZE, COLOR_DEFAULT);
+        if (len < 0) {
+            return;
+        }
+    }
+
+    fprintf(stderr, "%s", buf);
 }
 
+#define PRINT_INIT_ERR(level, err, log_fmt, log_args...)                                           \
+    output_before_doca_logger(level, "Initialization error: %s, %s. " log_fmt,                     \
+                              doca_error_get_name(err), doca_error_get_descr(err), ##log_args)
+
 void vlog_start(const char *log_module_name, vlog_levels_t log_level, const char *log_filename,
-                int log_details, bool log_in_colors)
+                bool log_in_colors)
 {
     g_vlogger_file = stderr;
 
-    g_vlogger_cb = xlio_log_get_cb_func();
-
     strncpy(g_vlogger_module_name, log_module_name, sizeof(g_vlogger_module_name) - 1);
     g_vlogger_module_name[sizeof(g_vlogger_module_name) - 1] = '\0';
 
-    vlog_get_usec_since_start();
-
     char local_log_filename[255];
     if (log_filename != NULL && strcmp(log_filename, "")) {
         sprintf(local_log_filename, "%s", log_filename);
         g_vlogger_fd = open(local_log_filename, O_WRONLY | O_CREAT | O_TRUNC, 0644);
         if (g_vlogger_fd < 0) {
-            vlog_printf(VLOG_PANIC, "Failed to open logfile: %s\n", local_log_filename);
-            exit(1);
+            __log_raw(VLOG_PANIC, "Failed to open logfile: %s\n", local_log_filename);
+            std::terminate();
```

Comment:
not sure exit(1) and std::terminate has the same behavior, lets discuss

Reply:
https://stackoverflow.com/questions/2820285/abort-terminate-or-exit

I think maybe we should throw exception and let the program end abnormally (instead of terminate)

but we can get back to the exit